diff --git a/src/H5Tinit_float.c b/src/H5Tinit_float.c
--- a/src/H5Tinit_float.c
+++ b/src/H5Tinit_float.c
@@ -608,8 +608,10 @@ H5T__init_native_float_types(void)
 #endif

 done:
+#ifdef FE_INVALID
     /* Clear any FE_INVALID exceptions from NaN handling */
     if (feclearexcept(FE_INVALID) != 0)
         HSYS_GOTO_ERROR(H5E_DATATYPE, H5E_CANTSET, FAIL, "can't clear floating-point exceptions");
+#endif

     /* Restore the original environment */
